pyat/at/physics/dynamic_aperture.py

```diff
@@ -0,0 +1,1008 @@
+        :param back_step:  number of steps to decrement to start with next search.
+        :return: 6D coordinates array of last surviving particle
+        """
+
+        # disable parallel computation
+        init_parcomp = self.parallel_computation
+        self.parallel_computation = False
+
+        # define initial coordinate
+        coord0 = {'x': 1e-6, 'xp': 0.0, 'y': 1e-6, 'yp': 0.0, 'delta': 0.0, 'ct': 0.0}
+
+        # define step in each plane
+        step = {pl: direction[ip]*(self.dict_def_range[pl][1] - self.dict_def_range[pl][0])
+                for ip, pl in enumerate(self.planes)}
+
+        limit = [v for _, v in coord0.items()]
+
+        for ir in range(number_of_recursions):
+
+            if self.verbose:
+                print('search {d} recursively step {s}/{ts}'.format(d=direction, s=ir, ts=number_of_recursions))
+
+            # define initial coordinates for search
+            #if ir > 0:
+            #    for pl in self.planes:
+            #        coord[pl] -= step[pl]  # go back by one of the previous size steps.
+            #else:
+            #    coord = copy.deepcopy(coord0)
+            if ir == 0:
+                coord = {'x': 1e-6, 'xp': 0.0, 'y': 1e-6, 'yp': 0.0, 'delta': 0.0, 'ct': 0.0}
+
+            if self.verbose:
+                print(coord)
+
+            # reduce step (initial = full range)
+            step = {k: v / self.search_divider for k, v in step.items()}
+
+            # do not recompute already computed point
+            if ir > 0:
+                for pl in self.planes:
+                    coord[pl] += step[pl]  # go back by one of the previous size steps.
+
+            # search limit
+            while self.test_survived(coord)[0]:
+                for pl in self.planes:
+                    coord[pl] += step[pl]
+                    # update tested points
+                    self.coordinates[pl].append(coord[pl])
+                if self.verbose:
+                    print('step forward')
+                    print(coord)
+
+                self.survived.append(True)
+
+            # last survived is previous point
+            for pl in self.planes:
+                _c = back_step * step[pl]
+                # limit to 0.0
+                if coord[pl] > 0.0:
+                    if coord[pl]-_c < 0.0:
+                        _c = coord[pl]
+                if coord[pl] < 0.0:
+                    if coord[pl]-_c > 0.0:
+                        _c = coord[pl]
+                # assign new start coord
+                coord[pl] -= _c
+
+            if self.verbose:
+                print('step back')
+                print(coord)
+
+            # last coordinates are lost and add an additional test point lost (for display and contour computation
+            if self.survived:
+                self.survived[-1] = False
+
+            coord_lost = copy.deepcopy(coord)
+            for pl in self.planes:
```

Comment:
Why 'deep' ? For a shallow copy, the `dict` object has a copy method more efficient that the generic copy. Better writing:

`coord_lost = coord.copy()`
or 
`coord_lost = dict(coord)`

Reply:
bacause I am scared not to get what I would get in matlab by this action. 
so to "save" the value of coord in coord_lost I used deepcopy. LAter on I change coord again, but I do not want coord_lost to change as well magically!

Reply:
So a shallow copy if that you need. Affecting a new valus to coord does not change coord__lost. Would you use a=deepcopy(42) ?

Reply:
I would! to make sure 42 is 42 and will be 42 also if b = a and b = 43!

Reply:
No ! Look:
```
a=object1
b=a
```
now if you **modify** `b`, like `append(b,42)`, of course `a` will be modified also. But if you assign something else to b:
`b=object2`
`a` will still pointing to `object1`, unmodified.

So for dicts:
```python
a['x'] = 0
b=a.copy()   # shallow copy, b is identical to a
b['x'] = 1   # a['x'] is unchanged
```
You assign 1 to b['x']. But a['x'] is still 0. That would be different you had put `b = a` instead of `b=a.copy()`.

A deep copy is a very complex operation, but is rarely necessary. And you never make copies of constants.